[XLA:GPU][ROCm] Refactor and extract platform dependent logic from cudnn_conv_algorithm_picker

jerryyin opened the pull request:
This PR creates a base class `gpu_conv_algorithm_picker` merges both `cudnn_conv_algorithm_picker` and `rocm_conv_algorithm_picker`. `PickBestAlgorithmNoCache()` function becomes platform dependent.

------

Test performed: //tensorflow/compiler/xla/tests:convolution_test all passed
@whchung @timshen91

jerryyin commented on tensorflow/compiler/xla/service/gpu/BUILD:
```diff
@@ -611,12 +617,12 @@ cc_library(
         "@com_google_absl//absl/strings:str_format",
         "@com_google_absl//absl/time",
         "@com_google_absl//absl/types:optional",
+        "//tensorflow/stream_executor/cuda:redzone_allocator",
```
FYI updated dependency location because `redzone_allocator` will become platform dependent and guarded by `if_cuda_is_configured` in the next PR.

jerryyin commented:
The check results failures are unrelated with this PR.

According to invocation log, the failure is due to 

> ERROR: /tmpfs/src/github/tensorflow/tensorflow/tools/api/golden/BUILD:3:1: invalid label '//tensorflow:tensorflow_py:__subpackages__' in element 0 of 'package' argument: invalid target name 'tensorflow_py:__subpackages__': target names may not contain ':'
ERROR: package contains errors: tensorflow/tools/api/golden

timshen91 commented:
@jerryyin , how much are the ROCM vs cuDNN implementations differ? The only difference I've seen is that ROCM doesn't have redzone checking / content checking yet. Otherwise, the portability should be already provided by AlgorithmDesc, ProfileResult, and RunCudnnConv (with a different name, of course). If we add a flag to turn off the checking, would that work for ROCM?

jerryyin commented:
@timshen91 Thanks for your review. You are right about redzone & comparator difference. Beside this, all difference comes from the implementation of the `PickBestAlgorithmNoCache()` function. Because it is coupled into the implementation from stream_executor's implementation. Assuming we want `ROCm` and `Cuda` into one class:

- We call the current implementation of convolution as "find mode". Regarding the find mode, we do not loop through and profile the convolution algorithm candidates. This means the implementation has to diverge between `ROCm` and `Cuda`.
- The on-going implementation of convolution is "immediate mode". Regarding the immediate mode, we are still experimenting between profiling/non-profiling versions. For the profiling version, the biggest difference come from the on going discussion from divergence of `GetAlgorithms()` divergence, from @deven-amd review [here](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/524/files#r298292150). Feel free to provide feedback on the new interface.
- Other than stuff mentioned above, we should be able to work through the platform related utility functions, like `CetCudnnVersion()`, `GetComputeCapability()`.

As a summary, the best chance to achieve unified implementation is dependent on:

- `ROCm` side implement and adopt Redzone allocator (This is not a prioritized feature, but on our list). It is probably not enough to add a flag to turn off Redzone checking from your side.
- The immediate mode choose to profile algorithms, as cuda do (dependent on @deven-amd's feature )
- The immediate mode does not use the new `GetAlgorithms()` inteface (dependent on [PR#524](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/524/files))

@whchung FYI

timshen91 commented:
> @timshen91 Thanks for your review. You are right about redzone & comparator difference. Beside this, all difference comes from the implementation of the `PickBestAlgorithmNoCache()` function. Because it is coupled into the implementation from stream_executor's implementation.

Thanks for the details. I'd suggest below:
* Keep the ROCM and Cuda code in the same class / file.
* Write `PickBestCudaAlgorithmNoCache` and `PickBestRocmAlgorithmNoCache` separately. Since `GetAlgorithms` only appears in `PickBestCudaAlgorithmNoCache`, and you feel free do whatever you want in `PickBestRocmAlgorithmNoCache`.

This way we allow (1) freedom for ROCM implementation, (2) reuse the caching code likewise, (3) and also keep the boilerplate down.

A while back I suggested the same approach for `core/kernel/conv_ops.cc` in https://github.com/tensorflow/tensorflow/pull/26018#issuecomment-476378148.

What do you think?

jerryyin commented:
@timshen91 I think that's reasonable to do, using inheritance can bring coupling problems that hard to resovle later. Let me make sure we are on the same page. Are you suggesting the following?

```
GpuConvAlgorithmPicker::PickBestCudaAlgorithmNoCache(){...}
GpuConvAlgorithmPicker::PickBestROCmAlgorithmNoCache(){...}
GpuConvAlgorithmPicker::PickBestAlgorithm(){
...
#if GOOGLE_CUDA
   PickBestCudaAlgorithmNoCache()
#elif TENSORFLOW_USE_ROCM
   PickBestROCmAlgorithmNoCache()
#endif

...
}
```
Do you agree with:

- Renaming `CudaConvAlgorithmPicker` to `GpuConvAlgorithmPicker`
- Use macros to branch between different implementations

timshen91 commented:
> @timshen91 I think that's reasonable to do, using inheritance can bring coupling problems that hard to resovle later. Let me make sure we are on the same page. Are you suggesting the following?
> 
> ```
> GpuConvAlgorithmPicker::PickBestCudaAlgorithmNoCache(){...}
> GpuConvAlgorithmPicker::PickBestROCmAlgorithmNoCache(){...}
> GpuConvAlgorithmPicker::PickBestAlgorithm(){
> ...
> #if GOOGLE_CUDA
>    PickBestCudaAlgorithmNoCache()
> #elif TENSORFLOW_USE_ROCM
>    PickBestROCmAlgorithmNoCache()
> #endif
> 
> ...
> }
> ```

Yes, except for the `#if`.

> 
> Do you agree with:
> 
> * Renaming `CudaConvAlgorithmPicker` to `GpuConvAlgorithmPicker`

Yes.

> * Use macros to branch between different implementations

No. I think `if (StreamExecutor::platform() == ...)` is a better branch, as it is a runtime branch and it allows both CUDA and ROCM logic to coexist - so that we don't recompile XLA for CUDA vs ROCM.

jerryyin commented:
@timshen91 Sounds good, will do

jerryyin committed:
Renaming cudnn_conv_algorithm_picker to gpu_conv_algorithm_picker

jerryyin committed:
Refactor Cuda implementation of pickBestAlgorithmNoCache()

jerryyin committed:
Adding ROCm implementation of pickBestAlgorithmNoCache()

jerryyin commented on tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc:
```diff
@@ -270,9 +272,10 @@ StatusOr<tensorflow::AutotuneResult>
-      absl::StrCat("GpuConvAlgorithmPicker::PickBestAlgorithmImpl for ",
-                   instr.ToString()));
+// Right now Redzone allocator is available in Cuda target only
+#if !TENSORFLOW_USE_ROCM
```
Note: This is a hack. Redzone allocator won't compile in ROCm side. I'm putting it inside the scope of this function so as to not to put #if guard everywhere.

timshen91 replied:
Maybe implement Redzone allocator with `return Unimplemented()` or `CHECK()`? `#if` in XLA won't work nicely internally, so let's try our best avoiding them.

jerryyin replied:
I can do that for redzone allocator, but that is not the only place where macro `#if` is used. I'll comment out two other cases FYI. Do you mean that `#if` will not work at all in your master branch?

timshen91 replied:
> Do you mean that #if will not work at all in XLA in master branch?

It will work in XLA in master branch, but I can't speak for our internal version (which is synchronized with the opensource version). Summoning @Artem-B for more details.

Empirically, if you search `tensorflow/compiler/xla` for `GOOGLE_CUDA`, you will find only one file that uses it and that file may be a no-op, as many XLA/GPU applications don't define `GOOGLE_CUDA`.

jerryyin replied:
Had a brief discussion offline with @whchung. I am planning to do the following:

-  Create a common interface in `tensorflow/stream_executor/redzone_allocator`, and make both `ROCm` and `Cuda` inherit from this interface and have their own implementation. The difference between `Cuda` and `ROCm` is in constructor, that in Cuda side, `cuda::PtxCompilationOptions` is passed to constructor. `ROCm` does not have this field, because `PtxCompilation` is `Cuda` only.
- Needs to update signature of `se::cuda::PtxCompilationOptions PtxOptsFromConfig()` to be able to return a Status. So that ROCm implementation can return unimplemented
- The real problem is this [line](https://github.com/tensorflow/tensorflow/blob/ece727fd8c8d278a32357702fddb66caf40d4572/tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc#L294). `se::cuda::RedzoneAllocator` cannot be visible to `ROCm` anyways, it looks like I still need a macro here to remove the `PtxOptsFromConfig(hlo_module_config)` field.  
Since you have clarified that the problem is more in `GOOGLE_CUDA` undefined in XLA side, it looks like it doesn't prevent us from using `TENSORFLOW_USE_ROCM`, which is defined in `ROCm` compilation no matter what. Do you agree to use `TENSORFLOW_USE_ROCM` just for this place then?

timshen91 replied:
It looks like we want a portable type called something like `GpuAsmOpts`, and it wraps `cuda::PtxCompilationOptions`. We've done that to GpuVersion and I don't mind doing the same thing here.

With that, we may not need an interface type as `redzone_allocator.h` may be fully portable?

> Since you have clarified that the problem is more in GOOGLE_CUDA undefined in XLA side, it looks like it doesn't prevent us from using TENSORFLOW_USE_ROCM, which is defined in ROCm compilation no matter what.

It's not really "no matter what". If `--config=rocm` is not passed in, TENSORFLOW_USE_ROCM won't be defined. Internally XLA/GPU works without `--config=cuda`.

jerryyin replied:
Thanks for the clarification in `--config=rocm` macro.

Regarding to `redzone_allocator.h`, I agree that it can be shared across `ROCm` and `Cuda`. For `Cuda`side, I can create a typedef inside `ptxas_utils` to make `GpuAsmOpts` and `cuda::PtxCompilationOptions` equivalent. `ROCm` side can be all empty returns "unimplemented" for now.

Artem-B replied:
TL;DR; I agree with @timshen91 

I'm missing a lot of context here, so my usefulness is rather limited here. I can provide a bit of historic context and a dose of personal opinion. Feel free to skip the rant below.

`GOOGLE_CUDA` has outlived its usefulness long ago, IMO. Remaining uses are often of the `happens to work, until it does not` kind and are not a good example to follow. There's fair amount of internal history behind it, the underlying causes for its existences (first iteration of CUDA compiler which was not quite CUDA-compatible) are gone now, but the clean ups didn't quite catch up to it and new use cases seem to be cargo-culted faster than cleanups remove them. :-(

XLA does build w/o --config=cuda, because it does not have any CUDA code to compile. It does need CUDA runtime to execute JIT'ed code, but that's just plain old C API of CUDA runtime. One exception is NCCL which we do want to use and which is one of the few CUDA-related libraries XLA uses that does require CUDA compilation. I'm not sure what's the current status, but I believe we were leaning pre-compiling it (which would use --config=cuda) and using it from XLA as a regular library, so we remain relatively weakly tied to CUDA. 

In this sense HIP runtime should follow the same pattern, IMO. If you're not compiling any HIP code for XLA, we can (and should, I think) strive to avoid strongly tying XLA to HIP, too. 

Current implementation may need refactoring to accommodate both runtimes, but I think long term it will result in better code overall. Hopefully it will get us to find better solutions for the issues we run into, as opposed of working around them in the most expedient way. 

Long term I want to get rid of --config=cuda altogether and just build CUDA-enabled targets along with CPU-only ones, so a single build could produce both at the same time. HIP would hopefully fit within the same model. Then it would be up to the user what to include in the executable and up to the stream executor how to launch the kernel on particular accelerator. That, however, assumes that if the code is shared between different accelerators, the checks should be done at runtime, and not at compile time. It may be OK to use compile-time defines in ROCm-specific file and, e.g. make a choice between a stub vs full implementation, but in common code compile-time checks will eventually become a problem. We already had first glimpses of that dealing with CPU/GPU variants. With more than one GPU and, likely, other accelerator variants on the horizon thing will only get more interesting.

jerryyin replied:
Awesome thank you for the background and history @Artem-B @timshen91. Feel free to always share your vision up here, those kind of information will make sure we are aligned in the long run. I have shared your comments to the `TF-ROCm` team. Personally speaking, it makes me have a clear understanding of how much we want to avoid those macros :-)

jerryyin commented on tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc:
```diff
@@ -479,15 +492,86 @@ StatusOr<AutotuneResult> CudnnConvAlgorithmPicker::PickBestAlgorithmNoCache(
+#if TENSORFLOW_USE_ROCM
+  // ROCm: Needs to set the first time caller flag
+  options.first_call_from_algorithm_picker = true;
+#endif
```
@timshen91 This is one place where in the find mode, the flag has to be set to make stream_executor happy. This flag exist in ROCm branch only.

jerryyin commented on tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc:
```diff
@@ -515,6 +599,11 @@ StatusOr<bool> CudnnConvAlgorithmPicker::RunOnInstruction(
+  // ROCm: Needs to set scratch size for ROCm stream executor make sure scratch
+  // space for convolution related calls are correctly allocated
+  backend_config.set_scratch_size(best_algo.scratch_bytes());
+#endif
```
@timshen91 This is another place where `set_scratch_size()` has to be called inside `rocm` build

timshen91 replied:
We already store the scratch bytes at line 521. XLA GPU convolution has two output, the result buffer and the scratch buffer. This is because XLA doesn't quite have a notion for in-out parameter, and we use output parameter for scratch buffer.

jerryyin replied:
For clarification, I was actually talking about adding the field in the end of `CudnnConvBackendConfig` [protobuf](https://github.com/tensorflow/tensorflow/blob/ece727fd8c8d278a32357702fddb66caf40d4572/tensorflow/compiler/xla/service/gpu/backend_configs.proto#L43). Which line 521 were you talking about?

timshen91 replied:
```
  Shape new_call_shape = ShapeUtil::MakeTupleShape(
      {instr->shape().tuple_shapes(0),
       ShapeUtil::MakeShape(U8, {best_algo.scratch_bytes()})});
```

It stores the scratch size as part of the output tuple shape.

jerryyin committed:
Address review feedback and implement redzone allocator in ROCm

jerryyin commented:
@timshen91 Addressed feedback in the newest commit. Thanks for your patience as this PR grows larger and larger. Brief explanation of the major changes:

- Moved `tensorflow/stream_executor/cuda/redzone_allocator.h` up one level, to make the header shared across `Cuda` and `ROCm`.
- Implement the boilerplate code for `ROCm` redzone_allocator.
- Moved PtxCompilationOptions from `tensorflow/stream_executor/cuda/ptxas_utils.h` to a independent header `tensorflow/stream_executor/gpu_asm_opts.h` and make the header shared across `Cuda` and `ROCm`.
- Manged to remove all `#if` macros from `gpu_conv_algorithm_picker`.
- and corresponding naming/dependency changes.

jerryyin commented on tensorflow/stream_executor/BUILD:
```diff
@@ -707,6 +715,17 @@ cc_library(
 )
 
+cc_library(
+    name = "redzone_allocator",
```
Desirably I want the `redzone_allocator` target to incorporate both `redzone_allocator_cuda` and `redzone_allocator_rocm`, but both implementation files are in child folders. I'm not sure if bazel allows that?

Besides, `if_cuda_is_configured` and `if_rocm_is_configured` is not yet available in this file. Will have to pull that utility in from `@local_config_rocm//rocm:build_defs.bzl` if we decided to go down this path.

jerryyin replied:
Thanks, will do.

jerryyin commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_runner.h:
```diff
@@ -35,6 +35,10 @@ struct RunConvOptions {
+
+  // Used by ROCm only: This is the first call to conv runner from
+  // the algorithm picker
+  bool first_call_from_algorithm_picker;
```
It is used in `cudnn_conv_runner.cc` as in [here](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/compiler/xla/service/gpu/cudnn_conv_runner.cc#L85). I will include it in the next commit.  There is a history to this. Briefly speaking, `ROCm` side convolution uses `find` mode. In the `find` mode, the first call has to be `miopenFind*` API. But the trigger for that is implicit and based on if `AlgorithmConfig::algorithm` is empty. 

It is undesirable such a coupling exists. However, `gpu_conv_algorithm_picker` was written with `Cuda` implementation in mind. The sequence of `GetAlgorithms()` then `RunCudnnConv()` does not fit for `ROCm` `find` mode.

Though, the good news is that `ROCm` `immediate` mode is almost there. Once `immediate` mode is out, we have a better chance to bring everything to parallel between `ROCm` and `Cuda`.

timshen91 replied:
> It is undesirable such a coupling exists. However, `gpu_conv_algorithm_picker` was written with `Cuda` implementation in mind. The sequence of `GetAlgorithms()` then `RunCudnnConv()` does not fit for `ROCm` `find` mode.

Maybe extract the `if (options.first_call_from_algorithm_picker) { ... }` logic out and put it in `PickBestAlgorithmNoCacheRocm()`?

jerryyin replied:
Sure, `first_call_from_algorithm_picker` does look wired, and I'm happy to get rid it when possible.

timshen91 commented on tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc:
```diff
@@ -254,7 +256,7 @@ StatusOr<AutotuneResult> GpuConvAlgorithmPicker::PickBestAlgorithm(
   StatusOr<AutotuneResult> result_or(InternalError("Unknown platform."));
   if (stream_exec_->platform_kind() == se::PlatformKind::kROCm) {
-    //result_or = PickBestAlgorithmNoCacheROCm(*instr, allocator, &stream);
+    result_or = PickBestAlgorithmNoCacheROCm(*instr, allocator, &stream);
```
Also, per Google style naming convention, it should be `PickBestAlgorithmNoCacheRocm`, not `PickBestAlgorithmNoCacheROCm`.

jerryyin replied:
Sounds good.

jerryyin commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_runner.cc:
```diff
@@ -80,7 +80,16 @@ Status RunCudnnConvImpl(const CudnnConvParams& params,
+  // empty. For all subsequent calls, we should use the value retrieved from
+  // the backend_config
+  if ((options.algo_override.has_value()) &&
+      (*options.algo_override == se::dnn::AlgorithmDesc())) {
```
@timshen91 FYI, I did this instead. When I take a second look at the interface, it does not make sense to expose the algorithm as a separate option for `ROCm` because the options argument already have one. So I create a empty algorithm as a signal to `ROCm` code base that this is the first time it is called.

jerryyin replied:
Also as an FYI, the best way to review the code changes now is to look directly at the files changed across all commits

timshen91 reviewed (dismissed):
The PR looks good to me now.

Thanks for the revises!

jerryyin commented:
@timshen91 Appreciate the effort in reviewing this large PR. It looks like there are some CI failures. I will make sure to get those fixed in the new commit. I will ping you once I'm done with it.

jerryyin committed:
Address bazel compilation errors on crossing subpackage boundary

jerryyin committed:
Address buildifier and fix Cuda compilation

jerryyin commented:
@timshen91 Would you mind take a second look? Since this PR touches a wide range of files, it is running into conflict often, and requires frequent rebase. Is there a way to have it merged quickly?

timshen91 commented:
> @timshen91 Would you mind take a second look? Since this PR touches a wide range of files, it is running into conflict often, and requires frequent rebase. Is there a way to have it merged quickly?

I'm trying to get it checked in. In general / next time, you can split one PR into multiple ones just for easy check-in. For example, this PR has a lot of renamings and non-functional refactoring that can be separated out.

jerryyin commented:
@timshen91 Thank you! Will remember to split it next time.

timshen91 commented on tensorflow/compiler/xla/service/gpu/BUILD:
```diff
@@ -1357,19 +1356,25 @@ tf_cc_test(
-    srcs = ["buffer_comparator.cc"],
+    srcs = [] + if_cuda_is_configured([
+        "buffer_comparator_cuda.cc",
+    ]) + if_rocm_is_configured([
```
This is also a compile-time dispatch that result in similar problems from `#ifdef`. Can you make buffer_comparator a runtime dispatch similar to what you did with redzone allocator?

jerryyin replied:
Thanks for the review. Let me try to understand the right way to do it:

The way I did with redzone allocator in the build file is that:
```
cc_library(
     name = "redzone_allocator",
     srcs = [] + if_rocm_is_configured([
         "//tensorflow/stream_executor/rocm:redzone_allocator.cc",
     ]) + if_cuda_is_configured([
         "//tensorflow/stream_executor/cuda:redzone_allocator.cc",
     ]),
     hdrs = ["redzone_allocator.h"],
     deps = [
     ...
```
It seems to be the exact same approach as in here, except that the component of redzone allocator is under stream executor (which I assume is the reason why it is not causing a problem from your end). 

Hip runtime compilation feature is not ready for us just yet, which is why I have been carefully not to implement related Cuda side runtime ptx compilation features, eg, `ptxas_utils.cc::CompilePtxOrGetCached()`. If the goal is to get rid of compile-time dispatch like `if_rocm_is_configured` from XLA, I can bring that part of interface parallel to ROCm as well.

cheshire replied:
#ifdef'ing the whole is worse than switching source at build time, which is worse than an isolated #ifdef, which is worse than virtual or runtime dispatch.
Hence the question: can we avoid build-time selection entirely? I understand that this would be a non-trivial change.

jerryyin replied:
Agreed on the sequence of ugliness :-). The rest of the thoughts sounds very ambitious if I understand it right. Based on the chain, switching the source at build time is worse than an isolated `#ifdef`. On a previous conversation I had with @timshen91, the impression I have is that the aim for XLA compilation is to compile without `--confg=rocm or cuda`. This makes us not having the option of doing any kind of `#ifdef`, and leaving the only option to runtime dispatch. (Let me know if there's any other way to do scoped `#ifdef`, independent from `--config` approach).

Following runtime dispatch thought, since `Cuda` and `ROCm` relying on different run-time libraries. To be able to get rid of build-time selection means that those run-time libraries has to exist on the same machine, or on a wired scenario `Cuda` run-time needs to exist on an `AMD` machine and vice versa. I'm not sure if this is acceptable. Even worse, because the whole code base needs to compile with both compilers this implicitly requires `HIP` to be able to compile with `Cuda` run-time. 

In the specific case, if I'm to get rid of `stream_executor`'s build-time selection. I am effectively exposing the ptx compilation and its very specific `Cuda` utilities to `ROCm` side of binary. It sounds like a bad idea.

If all the above concerns does not turn out to be a problem the run-time switching is not hard to achieve: For each device, try to call either `hipinit()` or `cuinit()`. Only one can succeed, and based on if the result is successful we can tell if this is on `ROCm` or on `Cuda`.

cheshire replied:
@jerryyin Do you want to discuss this over a call?

1) Removing #ifdefs out of TensorFlow proper is indeed a very large project which we may never finish. I'm talking about a much smaller scope: not having #ifdefs in stream executor and XLA.

> To be able to get rid of build-time selection means that those run-time libraries has to exist on the same machine

Why is this a problem? Those are just binaries we link to the final executable, right?

 >Even worse, because the whole code base needs to compile with both compilers this implicitly requires HIP to be able to compile with Cuda run-time.

StreamExecutor and XLA do not have CUDA kernels, so this should not be an issue.

> I am effectively exposing the ptx compilation and its very specific Cuda utilities to ROCm side of binary

Could you elaborate what precisely is bad? What kind of bugs do we potentially create this way?

jerryyin replied:
@cheshire Ah I've just seen your response this morning. In case we'd need further discussion, my email: zhuoran.yin@amd.com. The problem is mainly on compile time dependencies. For example, in [gpu_types.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/gpu/gpu_types.h#L45-L81), or [gpu_device.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/gpu_device.cc#L87-L101). I can hardly think of a way to get rid of those macros.

Taking one step back, this discussion is out of scope of the PR, which was aimed only for adding `ROCm` parallel logic to conv algorithm picker. In an offline discussion with @whchung, we agreed that this might be a good topic bring up from TF side so that we can talk about further collaborations. We also need to review the responsibilities between each team as well, e.g. bringing up the `Cuda` side of run-time may not be `ROCm` responsibility. Before this discussion happens, I can hardly think of a way to remove compile dispatch from `redzone_allocator`.

Now back to this single issue in `buffer_comparator`'s compile time dispatch. I can try to remove it by bring `stream_executor::ptxas_util` interface to parallel.

cheshire replied:
OK it's a good point that we have to limit the scope of this patch.

For a start, it seems easy enough to peel of the changes to RedzoneAllocator and BufferComparator?

Since those are currently no-ops for `ROCm`, it seems to me it would be best to:

 - Have a base class `RedzoneAllocator`
 - Have a method `static std::unique_ptr<RedzoneAllocator> GetAllocator()`
 - Have two subclasses: `NoopRedzoneAllocator` and `CudaRedzoneAllocator`
 - At runtime, dispatch to the second one if CUDA is available, and to the first one otherwise.
 - Perform the same steps for `BufferComparator`

@timshen91 @jerryyin Does it make sense? Any objections?

I've added quite a lot of other comments on this patch, but since it's so large it becomes hard to talk about it.
Once we limit the scope to convolutions only, it should get easier.

jerryyin replied:
@cheshire I'm planning on to start a new PR that includes only changes under stream executor. When that is done, come back to address compiler/xla related feedback.

First, I'm not sure if inheritance gives us anything more than the existing structure. Actually part of the lessen learned in exactly this PR is to avoid inheritance when possible, especially between `ROCm` and `Cuda` functionalities. This is the reason why it ended up in `PickBestAlgorithmNoCacheRocm()` and `PickBestAlgorithmNoCacheCuda()` instead of a inheritance structure. The problem with inheritance in active developing component is that when responsibility between child/parent class changed, a lot of case it will break `ROCm` build and requires constant refactoring.

Second for future extensiblility,  `ROCm` side plans on to actually implement the `RedzoneAllocator` as a mid-priority feature. From this perspective `NoopRedzoneAllocator` will be it a short-lived component.

Third, runtime dispatch is not as straightforward in the `redzone_allocator` component. The main roadblock comes from the divergence of `cuda/redzone_allocator.cc`. `cuda::redzone_allocator` take advantage of `ptxas_utils.cc`, which is dependent on `Cuda` runtime. For the minimal, the compile dispatch has to happen for `ptxas_utils`. Conceptually speaking, this is an allocator component, dispatching logic should stay outside of its responsibility.

My suggestion with the two components is to stay with the current implementation and we do not over-optimize. We can always refactor after `ROCm` side implementation for stream_executor is out, should we decide there is a lot of code duplication at that point, etc.

cheshire replied:
> First, I'm not sure if inheritance gives us anything more than the existing structure. 

Sure, it's up to you how exactly it should be done.

> Second for future extensiblility, ROCm side plans on to actually implement the RedzoneAllocator as a mid-priority feature

That's neither here nor there, but RedzoneAllocator was written for NVidia because it is closed source and does not support sanitization. It might make less sense for AMD.

I've mentioned `NoopRedzoneAllocator/BufferComparator` because in that case we would not need to add any extra compile-time guards apart from what we have, right?

jerryyin replied:
@cheshire Want to make sure we are on the same page now.

> RedzoneAllocator was written for NVidia because it is closed source and does not support sanitization. It might make less sense for AMD.

Good information. I noticed that the usage of RedzoneAllocator has been primarily related with convolution algorithms. Is it the case where you guys experienced most cudnn bugs,  in which the pattern is that cudnn writes more than it has claimed? or is it more related with we always have multiple candidates for convolution, and can discard problematic ones in the profiling round? Knowing this will help us make better decision on the form of this check. For example, it might make better sense to be a unit test in `MIOpen` rather than in TF runtime.

>  I've mentioned NoopRedzoneAllocator/BufferComparator because in that case we would not need to add any extra compile-time guards apart from what we have.

Does the existing review count as "extra" compile-time guards? I'm trying to understand what exactly changed from the first time this PR gets approved to it goes under re-review, possibly due to compile-time dispatch in XLA component? 

For the very minimal, I will have to use compile-time dispatch in `stream_executor::ptxas_utils.cc`. With it, I will be able to get rid of the compile time dispatch from `RedzoneAllocator` and `BufferComparator`. Would that approach work?

cheshire replied:
> Is it the case where you guys experienced most cudnn bugs, in which the pattern is that cudnn writes more than it has claimed

cheshire replied:
> Is it the case where you guys experienced most cudnn bugs, in which the pattern is that cudnn writes more than it has claimed? 

Yes, we have seen cudnn bugs with OOB writes.

> or is it more related with we always have multiple candidates for convolution, and can discard problematic ones in the profiling round

But once a convolution has done an OOB write we can't be 100% sure that it hasn't overwritten a crucial piece of memory.

> For example, it might make better sense to be a unit test in MIOpen rather than in TF runtime.

But we don't know which convolutions are problematic on which shapes in advance.
If we have open-source kernels ideally we should try to get asan working, but it would be a much larger investment.

> I'm trying to understand what exactly changed from the first time this PR gets approved to it goes under re-review, possibly due to compile-time dispatch in XLA component?

I'm grateful to you working on this, and I apologize for the delays from our side.
I think some decisions on what is the best way to do things are still being figured out, hence the discrepancy.

For now:

 - Yes, some amount of compile-time guards is unavoidable
 - Let's minimize/localize that as much as possible
 - Hence for BufferComparator/RedzoneAllocator you don't really need this, as ROCm version does not do anything, and having a noop version might be useful in any case.

jerryyin replied:
Ok, I think we are on the same page now. Thanks for your patience. I will come up with the PR that brings `ptxas_utils` to parallel to `ROCm`, and be the only component that does compile time dispatch. Once this is done, it will naturally remove the need to have a dummy implementation of `BufferComparator` or `RedzoneAllocator`. When my new PR is out, I will ping you and @timshen91 for review, as well as leaving a new message in this PR, and we can address any new issues from there.

jerryyin replied:
PR #31908 created, as well as addressed all review feedback there. I will rebase this PR once #31908 is merged.

cheshire commented on tensorflow/stream_executor/rocm/redzone_allocator.cc:
```diff
@@ -0,0 +1,71 @@
+}
+
+port::StatusOr<RedzoneCheckStatus> RedzoneAllocator::CheckRedzones() const {
+  return port::Status{port::error::UNIMPLEMENTED,
```
See comment above

cheshire commented on tensorflow/stream_executor/rocm/redzone_allocator.cc:
```diff
@@ -0,0 +1,71 @@
+}
+
+std::string RedzoneCheckStatus::RedzoneFailureMsg() const {
+  return absl::StrFormat("Redzone allocator is not implemented in ROCm");
```
`LOG(FATAL)` is more appropriate here, if this function is called, something went really wrong.

jerryyin replied:
Sounds good. Are you talking about add a `LOG(FATAL)` before the return?

cheshire commented on tensorflow/stream_executor/rocm/redzone_allocator.cc:
```diff
@@ -0,0 +1,71 @@
+                                   DeviceMemoryAllocator* memory_allocator,
+                                   GpuAsmOpts gpu_compilation_opts, int64 memory_limit,
+                                   int64 redzone_size, uint8 redzone_pattern)
+    : device_ordinal_(stream->parent()->device_ordinal()),
```
Should the constructor be doing anything?

jerryyin replied:
Probably not, will remove. Was including it because I thought `RoundUpToNearest` make sense.

cheshire commented on tensorflow/stream_executor/gpu_asm_opts.h:
```diff
@@ -0,0 +1,48 @@
+// Compilation options for compiling ptxas.
+struct GpuAsmOpts {
+  // Disable Cuda ptxas optimizations.
+  bool disable_ptxas_optimizations;
```
ptxas is CUDA-specific though. Should it be a more generic field name? What would be the AMD analogue?

jerryyin replied:
Hip run-time compilation feature is not available just yet. 

Judging the compilation pass:
`Cuda`: HLO - LLVM - PTX -cubin object
`ROCm`: HLO - LLVM - HSA cod object
There PTX doesn't have a parallel thing in `ROCm` side. Maybe `disable_asm_optimizations`?

cheshire commented on tensorflow/stream_executor/cuda/ptxas_utils.h:
```diff
@@ -19,31 +19,14 @@ limitations under the License.
-  }
-};
-
+using PtxCompilationOptions = GpuAsmOpts;
```
Is it possible to change the usage instead of creating an alias? It's pretty restricted.

jerryyin replied:
I can update everything to be `GpuAsmOpts`

cheshire commented on tensorflow/stream_executor/BUILD:
```diff
@@ -707,6 +723,30 @@ cc_library(
+    srcs = [] + if_rocm_is_configured([
+        "//tensorflow/stream_executor/rocm:redzone_allocator.cc",
+    ]) + if_cuda_is_configured([
+        "//tensorflow/stream_executor/cuda:redzone_allocator.cc",
```
Ideally, could we find a way to build both, and then choose the one to call using an `#ifdef` ?

jerryyin replied:
We can continue the discussion [here](https://github.com/tensorflow/tensorflow/pull/30884#discussion_r315300835)

cheshire commented on tensorflow/core/kernels/gpu_utils.cc:
```diff
@@ -19,14 +19,14 @@ limitations under the License.
 
-#include "google/protobuf/any.pb.h"
 #include "absl/algorithm/container.h"
+#include "google/protobuf/any.pb.h"
```
Blank line should separate tensorflow/ imports

jerryyin replied:
Sounds good.

cheshire commented on tensorflow/core/kernels/conv_ops_3d.cc:
```diff
@@ -450,8 +450,8 @@ struct LaunchConvOp<GPUDevice, T> {
-      se::cuda::RedzoneAllocator rz_allocator(
-          stream, &tf_allocator_adapter, se::cuda::PtxCompilationOptions());
+      se::RedzoneAllocator rz_allocator(stream, &tf_allocator_adapter,
+                                        se::cuda::PtxCompilationOptions());
```
GpuAsmOptions?

jerryyin replied:
Will get rid of `se::Cuda::PtxCompialtionOptions()` everywhere, this applies to similar cases below.

cheshire commented on tensorflow/core/kernels/conv_ops.cc:
```diff
@@ -955,18 +959,18 @@ void LaunchConv2DOp<GPUDevice, T>::operator()(
-          stream, &tf_allocator_adapter, se::cuda::PtxCompilationOptions(),
-          /*memory_limit=*/ConvolveScratchSize);
+      se::RedzoneAllocator rz_scratch_allocator(
+          stream, &tf_allocator_adapter, se::cuda::PtxCompilationOptions());
```
GpuAsmOptions?

cheshire commented on tensorflow/core/kernels/conv_ops.cc:
```diff
@@ -955,18 +959,18 @@ void LaunchConv2DOp<GPUDevice, T>::operator()(
-                                            se::cuda::PtxCompilationOptions());
+
+    se::RedzoneAllocator rz_allocator(stream, &tf_allocator_adapter,
+                                      se::cuda::PtxCompilationOptions());
```
GpuAsmOptions?

cheshire reviewed (changes requested):
Thanks for the patch! Sorry, I have a round of comments.

timshen91 commented:
Also, as this PR is growing larger, it's easier to split it into multiple PRs and check in separately. That way you don't have to rebase the whole PR all the time.

jerryyin commented:
@timshen91 Sounds good. Starting from next review, I will split it into two: One on XLA side gpu_conv_runner, another on stream executor side redzone_allocator. Will ping you when it is split into two.

jerryyin commented on tensorflow/compiler/xla/service/gpu/BUILD:
```diff
@@ -1357,19 +1356,25 @@ tf_cc_test(
 cc_library(
     name = "buffer_comparator",
-    srcs = ["buffer_comparator.cc"],
+    srcs = [] + if_cuda_is_configured([
```
Part of the discussion addressed in [above comments](https://github.com/tensorflow/tensorflow/pull/30884#discussion_r315300835). What I can do is to bring the `ROCm` stream executor side interface completely to parallel (to create a dummy interface but return `unimplemented error` for everything), so that both `Cuda` and `ROCm` implementation can be compiled and included in the same target. Then it is up to the runtime to choose the correct implementation based on platform id on the fly. 

Regarding the suggestion, I thought `#ifdef` is something we want to avoid in the code base? @timshen91 Just for clarification.

cheshire replied:
This is perhaps a most important point for this and similar CLs, so it's definitely worth discussing.

In terms of maintainability and complexity, surrounding everything with `#ifdef` is worse than switching source at build time, which is worse than an isolated `#ifdef`, which is worse than runtime dispatch (either virtual or an explicit switch).

I would prefer to avoid build-time tricks altogether, since they make the code less maintainable: it becomes not possible to build a binary capable of doing both, standard tooling stops working (since now #ifdef values have to be passed), getting build errors gets harder (as two build commands are required), etc etc.

jerryyin commented on tensorflow/compiler/xla/service/gpu/buffer_comparator_rocm.cc:
```diff
@@ -0,0 +1,52 @@
+    : shape_(shape), config_(config) {
+  // Normalize complex shapes: since we treat the passed array as a contiguous
+  // storage it does not matter which dimension are we doubling.
+  auto double_dim_size = [&]() {
```
Probably not, complex data type is not supported a lots of ROCm API calls. Will remove it.

jerryyin commented on tensorflow/core/kernels/conv_grad_filter_ops.cc:
```diff
@@ -55,7 +55,7 @@ limitations under the License.
 #if GOOGLE_CUDA
 #include "tensorflow/stream_executor/cuda/ptxas_utils.h"
-#include "tensorflow/stream_executor/cuda/redzone_allocator.h"
+#include "tensorflow/stream_executor/redzone_allocator.h"
```
That's right! In places where `#if` are still allowed, I'd prefer it to be included in the include-guard though, just so that one from `ROCm` side doesn't mistakenly used `redzone_allocator` thinking it is ready and only to figure out it is not when running.

jerryyin commented on tensorflow/stream_executor/rocm/redzone_allocator.cc:
```diff
@@ -0,0 +1,71 @@
+
+port::StatusOr<DeviceMemory<uint8>> RedzoneAllocator::AllocateBytes(
+    int64 byte_size) {
+  return port::Status{port::error::UNIMPLEMENTED,
```
Should I use XLA dependencies in stream executor?

jerryyin replied:
While addressing the review feedback I realized that the xla/util.h added the back-trace information with `WithLogBackTrace()`. I'm not going to pull over the convenient function here because:

1. I see no stream executor ever depends on an XLA component. It probably should be moved under core, but is out of scope of this PR.
2. Printing out the backtrace is nice but does not provide additional context to the specific case. 

I will mark similar comments as resolved.

mellanox-github commented:
Can one of the admins verify this patch?

cheshire commented on tensorflow/core/kernels/conv_ops.cc:
```diff
@@ -697,7 +697,11 @@ void LaunchConv2DOp<GPUDevice, T>::operator()(
       DataTypeToEnum<T>::value == DT_HALF && IsVoltaOrLater(*stream->parent());
+#else
+      false;
+#endif
```
Other convolution kernels in TF (conv_grad_filter_ops, conv_grad_input_ops) also have this variable defined in the same way. It seems this should be extracted into a function, and this change duplicated across those kernels?

cheshire commented on tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc:
```diff
@@ -516,11 +575,81 @@ StatusOr<AutotuneResult> CudnnConvAlgorithmPicker::PickBestAlgorithmNoCache(
+  options.algo_override = se::dnn::AlgorithmDesc();
+
+  bool launch_ok =
+      RunCudnnConv(&instr, absl::MakeSpan(operand_buffers), result_buffer,
```
Should PickBestAlgorithmNoCacheRocm be calling RunCudnnCovn?

jerryyin replied:
Everything in `cudnn_conv_runner.cc/h` should be renamed to `gpu_conv_runner` because both `ROCm` and and `Cuda` are sharing this utility. My newly submitted PR #32364 does only functionality changes. Renaming can come as another follow-up PR.

cheshire commented:
Should this PR be closed if it's being split?

cheshire commented:
Closing, since the PR was split.

jerryyin commented:
As an update. Both phase 1 #31908 (merged) and phase 2 #32364 (for review) of this PR are submitted.
